extensions/amp-story/0.1/media-pool.js

```diff
@@ -0,0 +1,748 @@
+      const attributeValue = fromAttributes[i].value;
+      if (!this.isProtectedAttributeName_(attributeName)) {
+        toEl.setAttribute(attributeName, attributeValue);
+      }
+    }
+  }
+
+  forEachMediaType_(callbackFn) {
+    Object.keys(MediaType).forEach(callbackFn.bind(this));
+  }
+
+
+  /**
+   * Invokes a function for all media managed by the media pool.
+   * @param {!function(!HTMLMediaElement)} callbackFn The function to be
+   *     invoked.
+   * @private
+   */
+  forEachMediaElement_(callbackFn) {
+    this.forEachMediaType_(key => {
+      const type = MediaType[key];
+      const allocatedEls = this.allocated[type];
+      allocatedEls.forEach(callbackFn.bind(this));
+    });
+
+    this.forEachMediaType_(key => {
+      const type = MediaType[key];
+      const unallocatedEls = this.unallocated[type];
+      unallocatedEls.forEach(callbackFn.bind(this));
+    });
+  }
+
+
+  /**
+   * Preloads the content of the specified media element in the DOM and returns
+   * a media element that can be used in its stead for playback.
+   * @param {!HTMLMediaElement} domMediaEl The media element, found in the DOM,
+   *     whose content should be loaded.
+   * @return {?HTMLMediaElement} A media element from the pool that can be used
+   *     to replace the specified element.
+   */
+  loadInternal_(domMediaEl) {
+    if (!isConnectedNode(domMediaEl)) {
+      // Don't handle nodes that aren't even in the document.
+      return null;
+    }
+
+    const mediaType = this.getMediaType_(domMediaEl);
+    if (this.getMatchingMediaElementFromPool_(mediaType, domMediaEl)) {
+      // The element being loaded is already an allocated media element.
+      return domMediaEl;
+    }
+
+    const sources = this.sources_[domMediaEl.id];
+    dev().assert(sources instanceof Sources,
+        `Cannot play unregistered element ${domMediaEl.outerHTML}.`);
+
+    const poolMediaEl = this.reserveUnallocatedMediaElement_(mediaType) ||
+        this.evictMediaElement_(mediaType, domMediaEl);
+
+    if (!poolMediaEl) {
+      // If there is no space in the pool to allocate a new element, and no
+      // element can be evicted, do not return any element.
+      return null;
+    }
+
+    this.swapPoolMediaElementIntoDom_(domMediaEl, poolMediaEl, sources);
+    this.allocateMediaElement_(mediaType, poolMediaEl);
+    return poolMediaEl;
+  }
+
+
+  /**
+   * "Blesses" the specified media element for playback.  In order for this to
+   * bless the media element, this function must be invoked in response to a
+   * user gesture.
+   * @param {!HTMLMediaElement} mediaEl The media element to bless.
+   * @return {!Promise} A promise that is resolved when blessing the media
+   *     element is complete.
+   */
+  bless_(mediaEl) {
+    const isPaused = mediaEl.paused;
+    const isMuted = mediaEl.muted;
+    const currentTime = mediaEl.currentTime;
+
+    mediaEl.muted = false;
+    return mediaEl.play().then(() => {
```

Comment:
`play` is not always a `promise` (MS Edge/IE)

Reply:
Wrapped in `Promise.resolve`, which (I think?) should handle both cases.